moveit_kinematics/test/test_kinematics_plugin.cpp

```diff
@@ -0,0 +1,730 @@
+  double timeout = 5.0;
+  moveit_msgs::MoveItErrorCodes error_code;
+  solution.resize(kinematics_solver_->getJointNames().size(), 0.0);
+  const std::vector<std::string>& fk_names = kinematics_solver_->getTipFrames();
+  robot_state::RobotState robot_state(robot_model_);
+
+  unsigned int success = 0;
+  for (unsigned int i = 0; i < num_ik_cb_tests_; ++i)
+  {
+    seed.resize(kinematics_solver_->getJointNames().size(), 0.0);
+    fk_values.resize(kinematics_solver_->getJointNames().size(), 0.0);
+    robot_state.setToRandomPositions(jmg_);
+    robot_state.copyJointGroupPositions(jmg_, fk_values);
+    std::vector<geometry_msgs::Pose> poses;
+    ASSERT_TRUE(kinematics_solver_->getPositionFK(fk_names, fk_values, poses));
+    if (poses[0].position.z <= 0.0f)
+      continue;
+
+    kinematics_solver_->searchPositionIK(poses[0], fk_values, timeout, solution,
+                                         boost::bind(&KinematicsTest::searchIKCallback, this, _1, _2, _3), error_code);
+    if (error_code.val == error_code.SUCCESS)
+      success++;
+    else
+      continue;
+
+    std::vector<geometry_msgs::Pose> reached_poses;
+    kinematics_solver_->getPositionFK(fk_names, solution, reached_poses);
+    EXPECT_NEAR_POSES(poses, reached_poses, IK_NEAR);
+  }
+
+  ROS_INFO_STREAM("Success Rate: " << (double)success / num_ik_cb_tests_);
+  EXPECT_GE(success, EXPECTED_SUCCESS_RATE * num_ik_cb_tests_);
+}
+
+TEST_F(KinematicsTest, getIK)
+{
+  std::vector<double> fk_values, solution;
+  moveit_msgs::MoveItErrorCodes error_code;
+  solution.resize(kinematics_solver_->getJointNames().size(), 0.0);
+  const std::vector<std::string>& fk_names = kinematics_solver_->getTipFrames();
+  robot_state::RobotState robot_state(robot_model_);
+
+  for (unsigned int i = 0; i < num_ik_tests_; ++i)
+  {
+    fk_values.resize(kinematics_solver_->getJointNames().size(), 0.0);
+    robot_state.setToRandomPositions(jmg_);
+    robot_state.copyJointGroupPositions(jmg_, fk_values);
+    std::vector<geometry_msgs::Pose> poses;
+
+    ASSERT_TRUE(kinematics_solver_->getPositionFK(fk_names, fk_values, poses));
+    kinematics_solver_->getPositionIK(poses[0], fk_values, solution, error_code);
+    // starting from the correct solution, should yield the same pose
+    EXPECT_EQ(error_code.val, error_code.SUCCESS);
+
+    Eigen::Map<Eigen::ArrayXd> sol(solution.data(), solution.size());
+    Eigen::Map<Eigen::ArrayXd> truth(fk_values.data(), fk_values.size());
+    EXPECT_TRUE(sol.isApprox(truth, IK_NEAR)) << sol.transpose() << std::endl << truth.transpose() << std::endl;
+  }
+}
+
+TEST_F(KinematicsTest, getIKMultipleSolutions)
+{
+  std::vector<double> seed, fk_values;
+  std::vector<std::vector<double>> solutions;
+  kinematics::KinematicsQueryOptions options;
+  kinematics::KinematicsResult result;
+
+  const std::vector<std::string>& fk_names = kinematics_solver_->getTipFrames();
+  robot_state::RobotState robot_state(robot_model_);
```

Comment:
should joints not part of jmg_ (if any) be initialized as well?

Reply:
Actually I was assuming that RobotState(RobotModel&) will initialize all joints to the default pose, but it doesn't. In this case, indeed, RobotState should be initialized correctly everywhere.

Reply:
We had quite a few crashes because it doesn't, we might actually want to change that in RobotState